R/git_tools.R

```diff
@@ -48,9 +48,47 @@ checkout_repo <- function(repo_dir, repo_url, select_branch_rule, token_envvar,
       message(paste("clone", repo_url, "to directory", repo_dir))
     }
 
-    git_repo <- git2r::clone(
-      url = repo_url, local_path = repo_dir, credentials = creds, progress = verbose >= 2
-    )
+    tryCatch({
+      git_repo <- git2r::clone(
+        url = repo_url, local_path = repo_dir, credentials = creds, progress = verbose >= 2
+      )
+    }, error = function(e) {
+      # catch some common errors
+      # only do this when cloning because the API calls introduce quite some time overhead
+      host <- paste(utils::head(strsplit(repo_url, "/", fixed = TRUE)[[1]], -2), collapse = "/")
+      repo <- paste(utils::tail(strsplit(repo_url, "/", fixed = TRUE)[[1]], 2), collapse = "/")
+      repo <- substr(repo, start = 0, stop = nchar(repo) - nchar(".git"))
+
+      if (!identical(httr::status_code(httr::HEAD(host)), 200L)) {
+        stop("Host ", host, " not reachable")
+      }
+
+      if (identical("https://github.com", host) || identical("https://code.roche.com", host)) {
+        # these queries should also work for Enterprise hosts
+        if (identical("https://github.com", host)) {
+          resp <- httr::GET(
+            paste0("https://api.github.com/repos/", repo),
+            # `token` argument not working
+            httr::add_headers(c(Authorization = paste("token", Sys.getenv(token_envvar))))
+          )
+        } else if (identical("https://code.roche.com", host)) {
+          resp <- httr::GET(
+            paste0(
+              "https://code.roche.com/api/v4/projects/",
+              utils::URLencode(repo, reserved = TRUE)
+            ),
+            httr::add_headers(c(Authorization = paste("Bearer", Sys.getenv(token_envvar))))
+          )
+        }
```

Comment:
It feels like the difference between the hosts belongs somewhere else rather than embedded in the code something like this - though not sure how much effort we want to spend on this especially if we are hard coding this for only 2 hosts
```
# some sort of configuration (could even be more sophisticated handling the tokens as well in the options?)
hosts <- list(`https://github.com`=list(request_fn = function(repo) httr::GET(...)), 
                    `https://code.roche.com` = list(request_fn = function(repo) ...)
                    
# then in code
if (host in names(hosts) && !is.null(hosts[[host]]$get_fn())){
  resp <- hosts[[host]]$get_fn()
  if(!identical(resp$status, ...) { ...
  }
}                    
```

Reply:
Yes, we should address this later, making it more general. Then, we need to say for each host whether it is Github or Gitlab.